src/library_dylink.js

```diff
@@ -960,12 +956,13 @@ var LibraryDylink = {
     // libName -> libData
     function loadLibData() {
       // for wasm, we can use fetch for async, but for fs mode we can only imitate it
-      if (flags.fs && flags.fs.findObject(libName)) {
-        var libData = flags.fs.readFile(libName, {encoding: 'binary'});
```

Comment:
This is a regression for us because we've been using the `fs` argument to `loadDynamicLibrary`. Do you have a recommended alternative @sbc100?

cc @ryanking13

Reply:
Can you call `_dlopen(...)` perhaps instead?

`loadDynamicLibrary` and the other symbols in `library_dylink.js` are not really part of the public API .. or at least I would prefer them not to be considered public.

Reply:
Not clear. What was the intended purpose of this hook if it's not supposed to be used downstream?

There's a few issues. One is asynchronous compilation: `dlopen` is synchronous I think, so it has to instantiate the module synchronously, which fails on the main thread for large modules.

A second problem is resolving dynamic library dependencies. If we `dlopen` a library `a.so` that depends on `b.so` then the lookup logic by default it does something we don't want.

Reply:
> Not clear. What was the intended purpose of this hook if it's not supposed to be used downstream?
> 
> There's a few issues. One is asynchronous compilation: `dlopen` is synchronous I think, so it has to instantiate the module synchronously, which fails on the main thread for large modules.

I think we can find a way to work around that.  There are 3 different versions of dlopen in emscripten and 2 of them are async (`emscripten_dlopen_promise` and `emscripten_dlopen`) .   They are mostly designed to be called from native code though.  There is you loading code happening?  Is it originating in native code?

> 
> A second problem is resolving dynamic library dependencies. If we `dlopen` a library `a.so` that depends on `b.so` then the lookup logic by default it does something we don't want.

Can you say little more about this.  The logic for loading the dependencies of dynamic libraries is all embedded in loadWebAssemblyModule on the JS side so I would expect them to be the same.

Reply:
> Not clear. What was the intended purpose of this hook if it's not supposed to be used downstream?

No, sadly we don't have much clarity about what is supposed to be internal vs public JS APIs.  I wish we did.   In this case we used that hook internally, but after this change it was no longer needed.   I'd rather not add it back because the new method of loading the data from the filesystem once on the main thread and sharing it via shared memory has a lot of advantages (the alternative is that each thread would need to do FS operations of their own, which themselves get proxied back to main thread, which adds a up to load wasted work).

Reply:
> Is it originating in native code?

No, the call originates from JavaScript, we have a function called `pyodide.loadPackage` which downloads and unpacks a bunch of zip files and preloads all the .so files in them. In the long run, we'd like to use JSPI to move away from having to do this preloading but for now all the loading is in JS.

> Can you say little more about this.

Well @ryanking13 understands this logic better than I do, but I will do my best. From reading our source code it seems that each Python package has a directory called something like `package_name/.libs` and they expect the dynamic library dependencies in the package to be resolved from the `.libs` directory. So the file layout would be:
```
package_name/a.so
package_name/.libs/b.so
```
and a.so has b.so as a dependency. So we somehow need dlopen to know to look in .libs for b.so.

`loadWebAssemblyModule`

Reply:
Perhaps if understood your use case a little more I could figure out what the a better solution might be.  Can you describe how/when/why you use your own code loading JS functions?

Reply:
Sorry .. we raced.  Thanks for the extra background.

Reply:
> the new method of loading the data from the filesystem once on the main thread and sharing it via shared memory has a lot of advantages

How is dynamic linking + pthreads support coming along? I haven't been following the developments since it seems that Pyodide's downstream struggles with COEP/COOP/CORP etc so it's unclear that using `Atomics` is a good route. But it would be cool to see that working!

Anyways since we're not using pthreads, there isn't as big of an issue...

Reply:
Are the functions in all these `.so` files at loaded via `dlsym`?   i.e. are you just preloading so that later `dlopen` doesn't block?  Or are the symbols in them required to even start the main program?  (e.g. are the dependencies coming from DT_NEEDED?)

Reply:
I think if we could set a search path that would be sufficient. I'm not sure how this works for native .so files, it would probably be helpful to figure that out.

Reply:
>  (e.g. are the dependencies coming from DT_NEEDED?)

I don't know. I'll try building with an instrumented version of `dynlink.c` and see what is happening.

Reply:
It seems like `dynlink.c` isn't compiled in, but it's the only place I see `DT_NEEDED`. How do I check for `DT_NEEDED`?

Reply:
I guess my question is: does the main program expect to be able to implicitly access the symbols in these dso files.. or does it access them specifically/individually via dlsym()?  (i.e. load time dynamic linking vs runtime dynamic linking?)

Reply:
i.e. can the main program start running at all without these DSO loaded?

Reply:
I think we're mostly using `dlsym`. The way Python C extensions work is in Python you say:
```py
import some_native_module
```
and it locates `some_native_module.so` and runs `dlopen("some_native_module.so")` then calls `dlsym("PyInit__some_native_module")`.

But it could be that `some_native_module.so` has been linked against `b.so` and needs the `b.so` symbols to run.

Reply:
Right, that makes sense.  So what you are trying to do here is pre-load the `.so` files so that dlopen/dlsym won't then later block.

In that case I think it makes sense to hook into the existing `preloadedWasm` mechanism.  Currently this is used by `FS_createPreloadedFile`  which then triggers the wasm preload plugin in `library_dylink.js`: https://github.com/emscripten-core/emscripten/blob/f655c30107258b61a0e81cb2f1b9e5b8292c65f1/src/library_dylink.js#L23-L41

`FS_createPreloadedFile` is a public API so maybe you could use that?

Reply:
We used to use `preloadedWasm`, but we transitioned away from it to our current approach because it fixed some subtle bug. So is the idea that we should:
1. locate so files we want to preload
2. Call `getDylinkMetadata` on each of them and resolve their dependent .so files using the search path
3. In topologically sorted order, call `FS_createPreloadedFile`
?

One related issue is we have to store global libs in `loadedLibsByName` twice: once with the absolute path and once just by name. I believe we'll also need to do this with `preloadedWasm`:
```js
if (global && Module.PATH.isAbs(lib)) {
  const libName: string = Module.PATH.basename(lib);
  const dso: any = Module.LDSO.loadedLibsByName[libName];
  if (!dso) {
    Module.LDSO.loadedLibsByName[libName] =
      Module.LDSO.loadedLibsByName[lib];
  }
}
```

Reply:
> We used to use `preloadedWasm`, but we transitioned away from it to our current approach because it fixed some subtle bug.

It would be interesting to know what that bug was, perhaps we can fix that?

Reply:
I would hope that whatever solution we come with we can avoid you having to call `getDylinkMetadata` directly.. thats also an API that I would consider internal to dynamic loader.

Reply:
I think the issue here is that we need to adjust how dependencies are looked up. If we call `FS.createPreloadedFile` directly, it will do its normal dependency resolution process without any way for us to control it. So we either need a hook or we need a way to work out the dependencies ahead of time and resolve them manually. As far as I can tell `getDylinkMetadata` is the only way for us to work out dependencies ahead of time.

Reply:
When you say "adjust how dependencies are looked up" do you mean controlling where .so files are found (like LD_LIBRARY_PATH)?     Or something else?

Is that because you have DSO with the same name in different locations?

Reply:
> do you mean controlling where .so files are found (like LD_LIBRARY_PATH)? Or something else?

This may be all we need, but I'm not 100% sure. Presumably there is some correct standard way to do this we could be using?

Reply:
Thanks for the ping. Let me make it a little clearer

> When you say "adjust how dependencies are looked up" do you mean controlling where .so files are found (like LD_LIBRARY_PATH)? Or something else?

Yes, "adjust how dependencies are looked up" is why Pyodide uses fs.findObject and fs.readfile now.

If a.so depends on b.so, we need a way to locate b.so somewhere in the (virtual) file system. So we want a global paths (LD_LIBRARY_PATH) that dependencies can be searched from.

Additionally, It would be great to have a thing like [RPATH](https://en.wikipedia.org/wiki/Rpath) (https://github.com/pyodide/pyodide/issues/3854). To make Python packages portable, we are trying to bundle shared libraries into Python wheel, so we want a way to search those "local" paths, which is availble in Linux with RPATH.

> Is that because you have DSO with the same name in different locations?

It's similar, but a little different.

The reason we modify `LDSO.loadedLibsByName` is to handle cases when we pass an absolute path and just a lib name into `loadDynamicLibrary`: when two libraries are the same.

For example, if we have a situation where we have previously called `loadDynamicLibrary (/usr/lib/b.so)` to load b.so, and we subsequently load `a.so` that depends on `b.so`, we want to know that the `b.so` that `a.so` depends on is `/usr/lib/b.so`, which has already been loaded, so that we don't get a duplicate load.

For now, `a.so` will call `loadDynamicLibrary(b.so)` in the process of loading the dependent library, so we need to tell it that this is the same file as `/usr/lib/b.so`.

Reply:
> As far as I can tell getDylinkMetadata is the only way for us to work out dependencies ahead of time.

This is not correct. The reason why we are using `getDylinkMetadata` now in Pyodide is because of the bug described in #18264.

If a library is loaded locally with RTLD_LOCAL option,
the dependent libraries will also be loaded locally. On Linux, dependent libraries can resolve symbols even if they are loaded locally, but I think there is currently a bug in Emscripten that prevents this.

As a workaround, Pyodide is changing the option to force libraries that are dependencies of other libraries to be loaded as RTLD_GLOBAL, using the metadata value returned by `getDylinkMetadata`.